File: frontend/src/lib/ipc/investigate.ts
import { Channel } from "@tauri-apps/api/core";

import { commands } from "./bindings";
import type {
  BlameChunk,
  BlameCommit,
  BlameSource,
  FileRevision,
  GitError,
  OriginEvent,
  OriginLine,
  OriginQuery,
  OriginReport,
  RepoId,
} from "./bindings";
import { CogitError } from "./index";

export type {
  BlameCommit,
  BlameSource,
  DeeperTarget,
  FileRevision,
  LineChange,
  LineMatch,
  Likelihood,
  OriginCandidate,
  OriginKind,
  OriginLine,
  OriginQuery,
  OriginReport,
  OriginText,
  PreviousFile,
} from "./bindings";

type Result<T> = { status: "ok"; data: T } | { status: "error"; error: GitError };

export interface BlameTables {
  commits: BlameCommit[];
  sources: BlameSource[];
  lines: OriginLine[];
}

/** Chunks may still be in flight when the command returns its count, so the promise
    settles on whichever of the two arrives last. */
function collect<T, C>(
  start: (channel: Channel<C>) => Promise<Result<number>>,
  take: (chunk: C) => void,
  size: () => number,
  done: () => T,
): Promise<T> {
  return new Promise((resolve, reject) => {
    let total: number | null = null;
    const settle = () => {
      if (total !== null && size() >= total) resolve(done());
    };
    const channel = new Channel<C>();
    channel.onmessage = (chunk) => {
      take(chunk);
      settle();
    };
    start(channel).then((result) => {
      if (result.status === "error") {
        reject(new CogitError(result.error));
        return;
      }
      total = result.data;
      settle();
    }, reject);
  });
}

export function investigateLog(
  repo: RepoId,
  path: string,
  rev: string | null,
  follow: boolean,
): Promise<FileRevision[]> {
  const rows: FileRevision[] = [];
  return collect<FileRevision[], FileRevision[]>(
    (channel) => commands.investigateLog(repo, path, rev, follow, channel),
    (chunk) => rows.push(...chunk),
    () => rows.length,
    () => rows,
  );
}

export function investigateBlame(
  repo: RepoId,
  path: string,
  rev: string | null,
  ignoreWhitespace: boolean,
): Promise<BlameTables> {
  const tables: BlameTables = { commits: [], sources: [], lines: [] };
  let header = false;
  return collect<BlameTables, BlameChunk>(
    (channel) => commands.investigateBlame(repo, path, rev, ignoreWhitespace, channel),
    (chunk) => {
      if (chunk.kind === "header") {
        header = true;
        tables.commits = chunk.commits;
        tables.sources = chunk.sources;
      } else {
        tables.lines.push(...chunk.lines);
      }
    },
    () => (header ? tables.lines.length : -1),
    () => tables,
  );
}

/** Resolves with the report, or `null` when the search was cancelled. `onStarted` hands
    over the id `cancelOriginSearch` needs. */
export function originCandidates(
  repo: RepoId,
  query: OriginQuery,
  onStarted: (id: number) => void,
): Promise<OriginReport | null> {
  return new Promise((resolve, reject) => {
    const channel = new Channel<OriginEvent>();
    channel.onmessage = (event) => {
      if (event.kind === "started") onStarted(event.id);
      else if (event.kind === "done") resolve(event.report);
      else resolve(null);
    };
    commands.originCandidates(repo, query, channel).then((result) => {
      if (result.status === "error") reject(new CogitError(result.error));
    }, reject);
  });
}

export async function cancelOriginSearch(id: number): Promise<void> {
  await commands.cancelOperation(id);
}

export async function openInvestigateWindow(url: string, title: string): Promise<void> {
  const result = await commands.openInvestigateWindow(url, title);
  if (result.status === "error") throw new CogitError(result.error);
}

File: src-tauri/src/commands/investigate.rs
//! The Investigate window (#15): the file's log, blame with origins, origin candidates.

use app_state::RepoId;
use git_engine::{
    BlameCommit, BlameSource, FileRevision, GitError, OriginLine, OriginQuery, OriginReport,
};
use serde::Serialize;

use super::blocking;

const CHUNK: usize = 200;

#[derive(Debug, Clone, Serialize, specta::Type)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum BlameChunk {
    /// Always first: the lines that follow index into these tables.
    Header {
        commits: Vec<BlameCommit>,
        sources: Vec<BlameSource>,
    },
    Lines {
        lines: Vec<OriginLine>,
    },
}

/// `Started` carries the id `cancel_operation` needs, long before the answer.
#[derive(Debug, Clone, Serialize, specta::Type)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum OriginEvent {
    Started { id: u32 },
    Done { report: OriginReport },
    Cancelled,
}

/// Every commit that changed the file, newest first, from `rev` (HEAD when absent).
#[tauri::command]
#[specta::specta]
pub async fn investigate_log(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    path: String,
    rev: Option<String>,
    follow: bool,
    on_chunk: tauri::ipc::Channel<Vec<FileRevision>>,
) -> Result<u32, GitError> {
    let app_state = state.state.clone();
    let log = blocking("investigate_log", move || {
        app_state.file_log(repo, &path, rev.as_deref(), follow)
    })
    .await?;
    for chunk in log.chunks(CHUNK) {
        let _ = on_chunk.send(chunk.to_vec());
    }
    Ok(u32::try_from(log.len()).unwrap_or(u32::MAX))
}

/// The file at `rev` (the working tree when absent), each line with its origin.
#[tauri::command]
#[specta::specta]
pub async fn investigate_blame(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    path: String,
    rev: Option<String>,
    ignore_whitespace: bool,
    on_chunk: tauri::ipc::Channel<BlameChunk>,
) -> Result<u32, GitError> {
    let app_state = state.state.clone();
    let started = std::time::Instant::now();
    let report = blocking("investigate_blame", move || {
        app_state.blame_origins(repo, &path, rev.as_deref(), ignore_whitespace)
    })
    .await?;
    tracing::debug!(
        repo = repo.0,
        lines = report.lines.len(),
        commits = report.commits.len(),
        elapsed_ms = started.elapsed().as_millis(),
        "blame with origins computed"
    );

    let total = u32::try_from(report.lines.len()).unwrap_or(u32::MAX);
    let _ = on_chunk.send(BlameChunk::Header {
        commits: report.commits,
        sources: report.sources,
    });
    for chunk in report.lines.chunks(CHUNK) {
        let _ = on_chunk.send(BlameChunk::Lines {
            lines: chunk.to_vec(),
        });
    }
    Ok(total)
}

/// Searches in the background where a block came from; `cancel_operation` stops it.
#[tauri::command]
#[specta::specta]
pub async fn origin_candidates(
    state: tauri::State<'_, crate::AppContext>,
    cancellations: tauri::State<'_, std::sync::Arc<crate::operations::Cancellations>>,
    repo: RepoId,
    query: OriginQuery,
    on_event: tauri::ipc::Channel<OriginEvent>,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    let cancellations = std::sync::Arc::clone(&cancellations);
    let (id, cancel) = cancellations.start();
    let _ = on_event.send(OriginEvent::Started { id });

    let started = std::time::Instant::now();
    let token = cancel.clone();
    let found = blocking("origin_candidates", move || {
        app_state.origin_candidates(repo, &query, &|| token.is_cancelled())
    })
    .await;
    cancellations.finish(id);

    let event = match found? {
        Some(report) => {
            tracing::debug!(
                candidates = report.candidates.len(),
                elapsed_ms = started.elapsed().as_millis(),
                "origin candidates found"
            );
            OriginEvent::Done { report }
        }
        None => OriginEvent::Cancelled,
    };
    let _ = on_event.send(event);
    Ok(())
}

/// `async`, unlike the older window commands: WebView2 can deadlock when a window is
/// built on the thread that delivered a synchronous command (R-283).
#[tauri::command]
#[specta::specta]
pub async fn open_investigate_window(
    app: tauri::AppHandle,
    url: String,
    title: String,
) -> Result<(), GitError> {
    crate::child_window::open(
        &app,
        "investigate",
        url,
        title,
        crate::child_window::Shape {
            width: 1280.0,
            height: 860.0,
            min_width: 820.0,
            min_height: 560.0,
        },
    )
    .map_err(|err| GitError::Internal(format!("cannot open the Investigate window: {err}")))
}
